### ISSUE TITLE
[VIX期限结构平坦化交易] | VIX Term Structure Flattening Trade

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4543]
**标签**: quantitative-strategy
### 摘要
基于VIX期限结构平坦化或倒挂现象进行均值回归交易，捕捉波动率曲面异常。
### 算法描述
**核心原理**：

VIX期限结构平坦化交易利用波动率曲面中近远月合约价差的均值回归特性。当VIX期限结构出现异常平坦（近月与远月VIX期货价差收窄）或倒挂时，策略预期这种状态不可持续，进行价差扩大方向的套利。

**数学模型**：

1. **VIX期限结构**：
   - 期货价格：$F(T) = S_0 \cdot e^{(r + \mu - \sigma^2/2)T}$
   - 期限结构斜率：$s = \frac{\partial F}{\partial T} = F(T) \cdot (r + \mu - \sigma^2/2)$
   - 平坦化指标：$\text{Flatness} = \frac{F(T_2) - F(T_1)}{T_2 - T_1}$

2. **Calendar Spread定价**：
   - 理论价差：$\text{Spread}_{theo} = F(T_2) \cdot (e^{\sigma^2 T_2/2} - e^{\sigma^2 T_1/2})$
   - 实际价差：$\text{Spread}_{actual} = F(T_2) - F(T_1)$
   - 价差比：$\text{Spread Ratio} = \frac{\text{Spread}_{actual}}{\text{Spread}_{theo}}$

3. **均值回归模型**：
   - Ornstein-Uhlenbeck过程：$dx = \theta(\mu - x)dt + \sigma dW$
   - 期望回归时间：$\mathbb{E}[\tau] = \frac{1}{\theta} \ln\left(\frac{x_0 - \mu}{x_0 - x_{target}}\right)$
   - 半衰期：$t_{1/2} = \frac{\ln 2}{\theta}$

4. **交易信号生成**：
   - Z-score：$Z = \frac{\text{Spread} - \mu_{\text{spread}}}{\sigma_{\text{spread}}}$
   - 入场阈值：$\text{Entry} = |Z| > Z_{\alpha}$
   - 离场阈值：$\text{Exit} = |Z| < Z_{\beta}$，其中 $Z_{\beta} < Z_{\alpha}$

5. **风险度量**：
   - VaR：$\text{VaR}_\alpha = \mu_{\text{loss}} - z_\alpha \sigma_{\text{loss}}$
   - 条件VaR：$\text{CVaR} = \mathbb{E}[\text{Loss} | \text{Loss} > \text{VaR}]$
   - 最大回撤：$\text{MDD} = \max_{t} \left(\max_{s \leq t} (P_s - P_t)\right)$

**策略流程**：
```
1. 获取VIX期货近月(1M)、次近月(2M)、季度(3M)合约价格
2. 计算期限结构斜率和平坦化指标
3. 计算Calendar Spread的Z-score
4. 当Z-score超过阈值时入场（买入低估合约，卖出高估合约）
5. 当价差回归均值或达到止损位时平仓
6. 管理每日mark-to-market风险
```

### Python代码示例
```python
import numpy as np
import pandas as pd
from scipy.stats import norm
from scipy.optimize import minimize_scalar, minimize
from datetime import datetime, timedelta
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
import warnings
warnings.filterwarnings('ignore')


@dataclass
class VIXFuturesQuote:
    """VIX期货报价"""
    symbol: str
    expiry_date: datetime
    futures_price: float
    term_month: int


@dataclass
class CalendarSpread:
    """日历价差"""
    near_symbol: str
    far_symbol: str
    near_price: float
    far_price: float
    spread: float
    theoretical_spread: float
    spread_ratio: float
    z_score: float
    signal: str


class VIXTermStructureModel:
    """VIX期限结构模型"""
    
    def __init__(
        self,
        mean_reversion_speed: float = 0.5,
        long_term_mean: float = 0.0,
        vol_of_vol: float = 0.02
    ):
        self.theta = mean_reversion_speed
        self.mu = long_term_mean
        self.sigma = vol_of_vol
    
    def ou_process_step(self, x: float, dt: float = 1/252) -> float:
        """Ornstein-Uhlenbeck过程单步模拟"""
        dx = self.theta * (self.mu - x) * dt + self.sigma * np.sqrt(dt) * np.random.normal()
        return x + dx
    
    def simulate_spread(self, x0: float, n_days: int = 60) -> np.ndarray:
        """模拟价差路径"""
        spreads = [x0]
        for _ in range(n_days):
            spreads.append(self.ou_process_step(spreads[-1]))
        return np.array(spreads)
    
    def expected_return_time(
        self,
        x0: float,
        target: float,
        dt: float = 1/252
    ) -> float:
        """计算期望回归时间"""
        if abs(x0 - self.mu) < 1e-6:
            return 0.0
        
        speed = abs(self.theta)
        if speed < 1e-6:
            return float('inf')
        
        half_life = np.log(2) / speed
        distance_ratio = abs(x0 - target) / abs(x0 - self.mu)
        
        return half_life * np.log(distance_ratio)


class VIXCalendarSpreadStrategy:
    """VIX期限结构平坦化交易策略"""
    
    def __init__(
        self,
        entry_threshold: float = 2.0,
        exit_threshold: float = 0.5,
        stop_loss: float = 3.0,
        position_size: float = 100000
    ):
        self.entry_threshold = entry_threshold
        self.exit_threshold = exit_threshold
        self.stop_loss = stop_loss
        self.position_size = position_size
        
        self.historical_spreads = []
        self.ou_model = VIXTermStructureModel()
        
        self.rolling_window = 60
    
    def calculate_term_structure_slope(
        self,
        futures: List[VIXFuturesQuote]
    ) -> float:
        """计算期限结构斜率"""
        if len(futures) < 2:
            return 0.0
        
        futures_sorted = sorted(futures, key=lambda x: x.expiry_date)
        
        price_changes = []
        time_diffs = []
        
        for i in range(1, len(futures_sorted)):
            dt = (futures_sorted[i].expiry_date - futures_sorted[0].expiry_date).days / 365.25
            dF = futures_sorted[i].futures_price - futures_sorted[0].futures_price
            
            if dt > 0:
                price_changes.append(dF)
                time_diffs.append(dt)
        
        if not time_diffs:
            return 0.0
        
        slope = np.average(np.array(price_changes) / np.array(time_diffs))
        
        return slope
    
    def calculate_flatness_index(
        self,
        near_price: float,
        far_price: float,
        near_expiry: datetime,
        far_expiry: datetime
    ) -> float:
        """计算平坦化指数"""
        time_to_near = (near_expiry - datetime.now()).days / 365.25
        time_to_far = (far_expiry - datetime.now()).days / 365.25
        
        if time_to_near <= 0 or time_to_far <= 0:
            return 0.0
        
        actual_spread = far_price - near_price
        
        implied_vol = 0.20
        
        theoretical_spread = near_price * (
            np.exp(implied_vol**2 * time_to_far / 2) -
            np.exp(implied_vol**2 * time_to_near / 2)
        )
        
        if theoretical_spread == 0:
            return 0.0
        
        flatness_index = actual_spread / theoretical_spread
        
        return flatness_index
    
    def calculate_z_score(self, spread: float) -> float:
        """计算Z-score"""
        if len(self.historical_spreads) < 20:
            return 0.0
        
        recent_spreads = self.historical_spreads[-self.rolling_window:]
        
        mu = np.mean(recent_spreads)
        sigma = np.std(recent_spreads)
        
        if sigma < 1e-6:
            return 0.0
        
        z = (spread - mu) / sigma
        
        return z
    
    def generate_signal(
        self,
        near_price: float,
        far_price: float,
        near_expiry: datetime,
        far_expiry: datetime
    ) -> Tuple[str, float, float]:
        """生成交易信号"""
        spread = far_price - near_price
        self.historical_spreads.append(spread)
        
        z_score = self.calculate_z_score(spread)
        
        flatness = self.calculate_flatness_index(
            near_price, far_price, near_expiry, far_expiry
        )
        
        if abs(z_score) > self.stop_loss:
            signal = "STOP_LOSS"
        elif abs(z_score) < self.exit_threshold:
            signal = "EXIT"
        elif z_score > self.entry_threshold:
            signal = "SHORT_SPREAD"
        elif z_score < -self.entry_threshold:
            signal = "LONG_SPREAD"
        else:
            signal = "NO_SIGNAL"
        
        return signal, z_score, flatness
    
    def calculate_position_pnl(
        self,
        near_price_entry: float,
        far_price_entry: float,
        near_price_current: float,
        far_price_current: float,
        position_type: str,
        notional: float
    ) -> float:
        """计算仓位盈亏"""
        near_pnl = (near_price_current - near_price_entry) / near_price_entry * notional
        far_pnl = (far_price_current - far_price_entry) / far_price_entry * notional
        
        if position_type == "LONG_SPREAD":
            return near_pnl - far_pnl
        elif position_type == "SHORT_SPREAD":
            return far_pnl - near_pnl
        else:
            return 0.0
    
    def calculate_var(
        self,
        returns: np.ndarray,
        confidence: float = 0.95
    ) -> float:
        """计算VaR"""
        if len(returns) < 2:
            return 0.0
        
        sorted_returns = np.sort(returns)
        idx = int((1 - confidence) * len(sorted_returns))
        
        return abs(sorted_returns[idx])
    
    def backtest_strategy(
        self,
        historical_data: pd.DataFrame,
        initial_capital: float = 1000000
    ) -> Dict:
        """回测策略"""
        capital = initial_capital
        position = None
        entry_spread = 0.0
        entry_prices = {}
        
        equity_curve = [initial_capital]
        drawdowns = [0.0]
        max_equity = initial_capital
        
        for i in range(1, len(historical_data)):
            date = historical_data.iloc[i]['date']
            near_price = historical_data.iloc[i]['near_price']
            far_price = historical_data.iloc[i]['far_price']
            near_expiry = datetime.strptime(str(historical_data.iloc[i]['near_expiry']), '%Y%m%d')
            far_expiry = datetime.strptime(str(historical_data.iloc[i]['far_expiry']), '%Y%m%d')
            
            signal, z_score, flatness = self.generate_signal(
                near_price, far_price, near_expiry, far_expiry
            )
            
            if position is None:
                if signal == "LONG_SPREAD":
                    position = "LONG_SPREAD"
                    entry_spread = far_price - near_price
                    entry_prices = {'near': near_price, 'far': far_price}
                elif signal == "SHORT_SPREAD":
                    position = "SHORT_SPREAD"
                    entry_spread = far_price - near_price
                    entry_prices = {'near': near_price, 'far': far_price}
            
            elif position is not None:
                pnl = self.calculate_position_pnl(
                    entry_prices['near'], entry_prices['far'],
                    near_price, far_price,
                    position, self.position_size
                )
                
                if signal in ["EXIT", "STOP_LOSS"] or z_score * (1 if position == "LONG_SPREAD" else -1) < -self.exit_threshold:
                    capital += pnl
                    position = None
                    entry_prices = {}
            
            equity = capital + pnl if position else capital
            equity_curve.append(equity)
            max_equity = max(max_equity, equity)
            drawdowns.append((max_equity - equity) / max_equity)
        
        returns = np.diff(equity_curve) / equity_curve[:-1]
        
        return {
            'final_capital': capital,
            'total_return': (capital - initial_capital) / initial_capital,
            'sharpe_ratio': np.mean(returns) / np.std(returns) * np.sqrt(252) if np.std(returns) > 0 else 0,
            'max_drawdown': max(drawdowns),
            'var_95': self.calculate_var(returns, 0.95),
            'equity_curve': equity_curve
        }


def simulate_vix_term_structure(n_days: int = 252) -> pd.DataFrame:
    """模拟VIX期限结构数据"""
    np.random.seed(42)
    
    dates = pd.date_range(start='2024-01-01', periods=n_days, freq='D')
    
    data = {
        'date': [],
        'near_price': [],
        'far_price': [],
        'near_expiry': [],
        'far_expiry': []
    }
    
    base_vix = 20.0
    
    for i, date in enumerate(dates):
        time_factor = np.sin(i / 30) * 2
        
        near_price = base_vix + time_factor + np.random.normal(0, 0.5)
        far_price = near_price + np.random.uniform(0.5, 2.0) + time_factor * 0.5
        
        near_expiry = (date + timedelta(days=30)).strftime('%Y%m%d')
        far_expiry = (date + timedelta(days=60)).strftime('%Y%m%d')
        
        data['date'].append(date)
        data['near_price'].append(near_price)
        data['far_price'].append(far_price)
        data['near_expiry'].append(near_expiry)
        data['far_expiry'].append(far_expiry)
    
    return pd.DataFrame(data)


if __name__ == "__main__":
    print("=" * 70)
    print("VIX期限结构平坦化交易策略")
    print("=" * 70)
    
    historical_data = simulate_vix_term_structure(252)
    
    strategy = VIXCalendarSpreadStrategy(
        entry_threshold=2.0,
        exit_threshold=0.5,
        stop_loss=3.0,
        position_size=100000
    )
    
    for i in range(20, len(historical_data)):
        near_price = historical_data.iloc[i]['near_price']
        far_price = historical_data.iloc[i]['far_price']
        near_expiry = datetime.strptime(str(historical_data.iloc[i]['near_expiry']), '%Y%m%d')
        far_expiry = datetime.strptime(str(historical_data.iloc[i]['far_expiry']), '%Y%m%d')
        
        signal, z_score, flatness = strategy.generate_signal(
            near_price, far_price, near_expiry, far_expiry
        )
        
        if signal != "NO_SIGNAL":
            print(f"Day {i}: Signal={signal}, Z-score={z_score:.2f}, Flatness={flatness:.2f}")
    
    results = strategy.backtest_strategy(historical_data, initial_capital=1000000)
    
    print("\n" + "=" * 70)
    print("回测结果")
    print("=" * 70)
    print(f"最终资金: ${results['final_capital']:,.2f}")
    print(f"总收益率: {results['total_return']:.2%}")
    print(f"夏普比率: {results['sharpe_ratio']:.2f}")
    print(f"最大回撤: {results['max_drawdown']:.2%}")
    print(f"95% VaR: ${results['var_95'] * 1000000:,.2f}")
```

### 数据来源
- Reddit: r/quant, r/algorithmictrading
- 小红书: 量化投资
- 知乎: 量化交易
